src/dataspaces.jl

```diff
@@ -121,51 +121,165 @@ function isnull(dspace::Dataspace)
+    start0, stride, count, block = API.h5s_get_regular_hyperslab(dspace)
+    N = length(start0)
+    ntuple(N) do i
+        ri = N - i + 1
+        @inbounds BlockRange(start0[ri], stride[ri], count[ri], block[ri])
     end
-    dsel_id     = API.h5s_copy(dspace)
-    dsel_start  = Vector{API.hsize_t}(undef, n_dims)
-    dsel_stride = Vector{API.hsize_t}(undef, n_dims)
-    dsel_count  = Vector{API.hsize_t}(undef, n_dims)
-    for k in 1:n_dims
-        index = I[n_dims - k + 1]
-        if isa(index, Integer)
-            dsel_start[k] = index - 1
-            dsel_stride[k] = 1
-            dsel_count[k] = 1
-        elseif isa(index, AbstractRange)
-            dsel_start[k] = first(index) - 1
-            dsel_stride[k] = step(index)
-            dsel_count[k] = length(index)
+end
+
+struct BlockRange
+    start0::API.hsize_t
+    stride::API.hsize_t
+    count::API.hsize_t
+    block::API.hsize_t
+end
+
+"""
+    HDF5.BlockRange(;start::Integer, stride::Integer=1, count::Integer=1, block::Integer=1)
+
+A `BlockRange` represents a selection along a single dimension of a HDF5
+hyperslab. It is similar to a Julia `range` object, with some extra features for
+selecting multiple contiguous blocks.
+
+- `start`: the index of the first element in the first block (1-based).
+- `stride`: the step between the first element of each block (must be >0)
+- `count`: the number of blocks (can be -1 for an unlimited number of blocks)
+- `block`: the number of elements in each block.
+
+    BlockRange(obj::Union{Integer, OrdinalRange})
+
+Convert `obj` to a `BlockRange` object.
+
+# External links
+- [HDF5 User Guide, section 7.4.2.1 "Selecting Hyperslabs"](https://support.hdfgroup.org/HDF5/doc/UG/HDF5_Users_Guide-Responsive%20HTML5/index.html#t=HDF5_Users_Guide%2FDataspaces%2FHDF5_Dataspaces_and_Partial_I_O.htm%23TOC_7_4_2_Programming_Modelbc-8&rhtocid=7.2.0_2)
+"""
+function BlockRange(; start::Integer, stride::Integer=1, count::Integer=1, block::Integer=1)
+    if count == -1
+        count = API.H5S_UNLIMITED
+    end
+    BlockRange(start - 1, stride, count, block)
+end
+BlockRange(start::Integer; stride=1, count=1, block=1) =
+    BlockRange(; start=start, stride=stride, count=count, block=block)
+BlockRange(r::AbstractUnitRange; stride=max(length(r), 1), count=1) =
+    BlockRange(; start=first(r), stride=stride, count=count, block=length(r))
+BlockRange(r::OrdinalRange) = BlockRange(; start=first(r), stride=step(r), count=length(r))
+BlockRange(br::BlockRange) = br
+
+Base.to_index(d::Dataset, br::BlockRange) = br
+Base.length(br::BlockRange) = Int(br.count * br.block)
+
+function Base.show(io::IO, br::BlockRange)
+    start = Int(br.start0 + 1)
+    # choose the simplest possible representation
+    if br.count == 1
+        if br.block == 1
+            # integer
+            print(io, start)
         else
-            error("index must be range or integer")
+            # UnitRange
+            print(io, range(start; length=Int(br.block)))
```

Comment:
```suggestion
            print(io, "BlockRange(")
            print(io, range(start; length=Int(br.block)))
            print(io, ")")
```

Reply:
This will change the printing of Dataspaces, but perhaps we could make that a `compact` option?

Reply:
I've added this.